File: 21. Ch-5-4.py
class TwoSum:
    def find_indices(self, arr, target):
        hashmap = {}
        print(f"Finding two indices whose values add up to {target}...")
        for i in range(len(arr)):
            x=arr[i]
            y = target - x
            if y in hashmap:
                print(f"Indices: {hashmap[y]} {i}")
                return
            hashmap[arr[i]] = i
        print("No solution found")

n = int(input().strip())
arr = list(map(int, input().split()))
target = int(input().strip())

finder = TwoSum()
finder.find_indices(arr, target)

File: 21. Ch-5-4.java
import java.util.*;
class TwoSum 
{
    public void findTwoSum(int[] arr, int target) 
    {
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < arr.length; i++) 
        {
            int required = target - arr[i];
            if (map.containsKey(required)) 
            {
                System.out.println("Finding two indices whose values add up to " + target + "...");
                System.out.println("Indices: " + map.get(required) + " " + i);
                return;
            }
            map.put(arr[i], i);
        }
        System.out.println("No solution found");
    }
}

public class Main 
{
    public static void main(String[] args) 
    {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }
        int target = sc.nextInt();

        TwoSum obj = new TwoSum();
        obj.findTwoSum(arr, target);
    }
}
